File: 127.word-ladder.js
/*
 * @lc app=leetcode id=127 lang=javascript
 *
 * [127] Word Ladder
 */

// @lc code=start
/**
 * @param {string} beginWord
 * @param {string} endWord
 * @param {string[]} wordList
 * @return {number}
 */
var ladderLength = function (beginWord, endWord, wordList) {
  let wordSet = new Set(wordList);
  if (!wordSet.has(endWord)) return 0;

  let replaceAt = (s, i, c) => {
    return s.slice(0, i) + c + s.slice(i + 1);
  };

  let queue = [beginWord];
  let res = 0;

  while (queue.length > 0) {
    for (let q = queue.length; q > 0; --q) {
      let word = queue.shift();
      if (word === endWord) return res + 1;
      for (let i = 0; i < word.length; ++i) {
        for (let ch = 97; ch < 123; ++ch) {
          let w = replaceAt(word, i, String.fromCharCode(ch));
          if (wordSet.has(w) && w !== word) {
            queue.push(w);
            wordSet.delete(w);
          }
        }
      }
    }
    ++res;
  }

  return 0;
};
// @lc code=end

File: 127.word-ladder.py
#
# @lc app=leetcode id=127 lang=python3
#
# [127] Word Ladder
#

# bidirectional bfs
# Time:  O(b ^ (d / 2)), b is the branch factor of bfs, d is the result depth
# Space: O(w * l), w is the number of words, l is the max length of words

# bfs
# Time:  O(b ^ d)
# Space: O(w * l)

import collections
import string
from typing import List


# @lc code=start
class Solution:
    def ladderLength(self, beginWord: str, endWord: str, wordList: List[str]) -> int:
        words = set(wordList)
        if endWord not in words:
            return 0
        left, right = {beginWord}, {endWord}
        res = 2
        while left:
            words -= left
            new_left = set()
            for word in left:
                for i in range(len(word)):
                    for j in string.ascii_lowercase:
                        new_word = word[:i] + j + word[i + 1 :]
                        if new_word in right:
                            return res
                        if new_word in words:
                            new_left.add(new_word)
            left = new_left
            res += 1
            if len(left) > len(right):
                left, right = right, left
        return 0

        words = set(wordList)
        if endWord not in words:
            return 0
        res, q = 2, collections.deque([beginWord])
        while q:
            q_len = len(q)
            for _ in range(q_len):
                word = q.popleft()
                for i in range(len(word)):
                    for j in string.ascii_lowercase:
                        new_word = word[:i] + j + word[i + 1 :]
                        if new_word == endWord:
                            return res
                        if new_word in words:
                            words.remove(new_word)
                            q.append(new_word)
            res += 1
        return 0


# @lc code=end
